### PR TITLE
Remove SSH-based drivers from ironic (bsc#1087472)

### PR DESCRIPTION
**Pull request by sayalilunkad**
Removing all SSH based drivers from config file and added validaion from ironic
proposal to not allow these drivers.

### CODE REVIEW

File: chef/cookbooks/ironic/recipes/server.rb [hunk 1]
```diff
@@ -54,9 +54,13 @@
 end
 
 node[:ironic][:enabled_drivers].each do |d|
-  driver_dependencies = node[:ironic][:platform][:driver_dependencies][d] || []
-  driver_dependencies.each do |p|
-    package p
+  if d.include?("ssh")
+    node[:ironic][:enabled_drivers].delete(d)
```
Comment (toabctl) [inline chef/cookbooks/ironic/recipes/server.rb]: Not sure, but shouldn't that be done via a crowbar migration?

Comment (toabctl) [reply chef/cookbooks/ironic/recipes/server.rb]: Yes. But don't we want to delete these drivers also from the proposal? My understanding is, that theses drivers are no longer valid so they should not stay in the crowbar ironic proposal.

Comment (skazi0) [reply chef/cookbooks/ironic/recipes/server.rb]: I was going to suggest the same. The question is: would it only fix the proposal or also node attributes (i.e. would users need to apply the proposal)? Maybe both would be needed?
@sayalilunkad if you insist on doing it in the cookbook, I would copy `node[:ironic][:enabled_drivers]` to a local variable, filter that and use it in the following places to make it clear that you don't modify the node attributes.

Comment (sayalilunkad) [reply chef/cookbooks/ironic/recipes/server.rb]: @toabctl @skazi0 yes that makes sense that we remove it from the proposal. I think this should update the node attribute as well but I will check.

File: chef/data_bags/crowbar/migrate/ironic/204_remove_ssh_based_drivers.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+def upgrade(ta, td, a, d)
+  a["ironic"]["enabled_drivers"] = ta["ironic"]["enabled_drivers"]
+  a["ironic"]["enabled_drivers"].each do |driver|
+    a["ironic"]["enabled_drivers"].delete(driver) if driver.include?("ssh")
+  end
+  return a, d
+end
+
+def downgrade(ta, td, a, d)
+  a["ironic"]["enabled_drivers"] = ta["ironic"]["enabled_drivers"]
```
Comment (skazi0) [inline chef/data_bags/crowbar/migrate/ironic/204_remove_ssh_based_drivers.rb]: I don't think we should do anything in downgrade.

File: chef/data_bags/crowbar/migrate/ironic/204_remove_ssh_based_drivers.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+def upgrade(ta, td, a, d)
+  a["ironic"]["enabled_drivers"] = ta["ironic"]["enabled_drivers"]
```
Comment (skazi0) [inline chef/data_bags/crowbar/migrate/ironic/204_remove_ssh_based_drivers.rb]: Why overwrite user's drivers with default?

File: chef/data_bags/crowbar/migrate/ironic/204_remove_ssh_based_drivers.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  a["ironic"]["enabled_drivers"].each do |driver|
```
Comment (skazi0) [inline chef/data_bags/crowbar/migrate/ironic/204_remove_ssh_based_drivers.rb]: Maybe something like this?
`a["ironic"]["enabled_drivers"] = a["ironic"]["enabled_drivers"].reject { |d| d.include? "ssh" }`

File: chef/cookbooks/ironic/attributes/default.rb [hunk 5]
```diff
@@ -28,10 +28,6 @@
       "openstack-ironic-api",
       "openstack-ironic-conductor"
     ],
-    driver_dependencies: {
```
[the remark this answers is not in the document]
Comment (sayalilunkad) [reply chef/cookbooks/ironic/attributes/default.rb]: Leaving a empty hastable for now incase there are other dependencies in the future.

File: crowbar_framework/app/models/ironic_service.rb [hunk 6]
```diff
@@ -94,6 +94,10 @@ def validate_proposal_after_save(proposal)
       validation_error I18n.t("barclamp.#{@bc_name}.validation.no_drivers")
     end
 
+    if proposal["attributes"][@bc_name]["enabled_drivers"].any? { |d| d.include?("ssh") }
```
[the remark this answers is not in the document]
Comment (sayalilunkad) [reply crowbar_framework/app/models/ironic_service.rb]: @skazi0 I agree.

File: chef/data_bags/crowbar/migrate/ironic/204_remove_ssh_based_drivers.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+def upgrade(ta, td, a, d)
+  a["ironic"]["enabled_drivers"] =
+    a["ironic"]["enabled_drivers"].reject { |driver| driver.include?("ssh") }
+  return a, d
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/ironic/204_remove_ssh_based_drivers.rb]: Did you test the downgrade? Does it work if the function is not defined? Maybe we need an empty function?

File: crowbar_framework/app/models/ironic_service.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (toabctl) [reply crowbar_framework/app/models/ironic_service.rb]: oki. thanks for clarification

**Pull request merged**